File: android2_lesson03/src/java/java/com/example/ama/android2_lesson03/repo/data/PreferencesMarkerManager.java
package com.example.ama.android2_lesson03.repo.data;

import android.content.Context;
import android.content.SharedPreferences;

import com.example.ama.android2_lesson03.PocketMap;
import com.example.ama.android2_lesson03.repo.data.base.MarkerManager;
import com.example.ama.android2_lesson03.repo.model.SimpleMarker;
import com.google.android.gms.maps.model.LatLng;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

/**
 * Class for managing favorite markers, working with SharedPreferences
 */
public class PreferencesMarkerManager implements MarkerManager {
    private static final String PREFERENCES_FILE_NAME = "marker_preferences";
    private static final String MARKERS_KEY = "all_markers";

    private static final String MARKER_TITLE_KEY = "title";
    private static final String MARKER_LATITUDE_KEY = "latitude";
    private static final String MARKER_LONGITUDE_KEY = "longitude";

    private SharedPreferences preferences;

    public PreferencesMarkerManager() {
        preferences = PocketMap.getInstance().getSharedPreferences(PREFERENCES_FILE_NAME, Context.MODE_PRIVATE);
    }

    @Override
    public ArrayList<SimpleMarker> getAllMarkers() {
        Set<String> markerSet = getFromPreferences();
        ArrayList<SimpleMarker> markers = new ArrayList<>();
        for (String o : markerSet) {
            markers.add(simpleMarkerFromJson(o));
        }
        return markers;
    }

    @Override
    public void addMarker(SimpleMarker marker) {
        Set<String> markerSet = getFromPreferences();
        markerSet.add(simpleMarkerToJson(marker));
        putInPreferences(markerSet);
    }

    @Override
    public void updateMarker(SimpleMarker marker, String newName) {
        Set<String> markerSet = getFromPreferences();
        markerSet.remove(simpleMarkerToJson(marker));
        marker.setTitle(newName);
        markerSet.add(simpleMarkerToJson(marker));
        putInPreferences(markerSet);
    }

    @Override
    public void deleteMarker(SimpleMarker marker) {
        Set<String> markerSet = getFromPreferences();
        markerSet.remove(simpleMarkerToJson(marker));
        putInPreferences(markerSet);
    }

    private void putInPreferences(Set<String> markerSet) {
        preferences.edit().clear().putStringSet(MARKERS_KEY, markerSet).apply();
    }

    private Set<String> getFromPreferences() {
        return preferences.getStringSet(MARKERS_KEY, new HashSet<String>());
    }


    private SimpleMarker simpleMarkerFromJson(String json) {
        try {
            JSONObject jsonObject = new JSONObject(json);
            return new SimpleMarker(
                    jsonObject.getString(MARKER_TITLE_KEY),
                    new LatLng(jsonObject.getDouble(MARKER_LATITUDE_KEY), jsonObject.getDouble(MARKER_LONGITUDE_KEY)));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;

    }

    private String simpleMarkerToJson(SimpleMarker marker) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put(MARKER_TITLE_KEY, marker.getTitle());
            jsonObject.put(MARKER_LATITUDE_KEY, marker.getPosition().latitude);
            jsonObject.put(MARKER_LONGITUDE_KEY, marker.getPosition().longitude);
            return jsonObject.toString();
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: android2_lesson03/src/kotlin/java/com/example/ama/android2_lesson03/repo/data/PreferencesMarkerManager.kt
package com.example.ama.android2_lesson03.repo.data

import android.content.Context
import com.example.ama.android2_lesson03.PocketMap
import com.example.ama.android2_lesson03.repo.data.base.MarkerManager
import com.example.ama.android2_lesson03.repo.model.SimpleMarker
import com.google.android.gms.maps.model.LatLng
import org.json.JSONException
import org.json.JSONObject

/**
 * Class for managing favorite markers, working with SharedPreferences
 */
class PreferencesMarkerManager : MarkerManager {

    companion object {
        const val PREFERENCES_FILE_NAME = "marker_preferences"
        const val MARKERS_KEY = "all_markers"

        const val MARKER_TITLE_KEY = "title"
        const val MARKER_LATITUDE_KEY = "latitude"
        const val MARKER_LONGITUDE_KEY = "longitude"
    }

    private val preferences = PocketMap.instance.getSharedPreferences(PREFERENCES_FILE_NAME, Context.MODE_PRIVATE)

    override fun getAllMarkers(): ArrayList<SimpleMarker> =
            getFromPreferences().mapNotNull { it -> simpleMarkerFromJson(it) } as ArrayList<SimpleMarker>


    override fun addMarker(marker: SimpleMarker) {
        val markerSet = getFromPreferences() as HashSet
        markerSet.add(simpleMarkerToJson(marker)!!)
        putInPreferences(markerSet)
    }

    override fun deleteMarker(marker: SimpleMarker) {
        val markerSet = getFromPreferences() as HashSet
        markerSet.remove(simpleMarkerToJson(marker)!!)
        putInPreferences(markerSet)
    }

    override fun updateMarker(marker: SimpleMarker, newName: String) {
        val markerSet = getFromPreferences() as HashSet
        markerSet.remove(simpleMarkerToJson(marker)!!)
        marker.title = newName
        markerSet.add(simpleMarkerToJson(marker)!!)
        putInPreferences(markerSet)
    }

    private fun getFromPreferences(): Set<String> =
            preferences.getStringSet(MARKERS_KEY, HashSet<String>())

    private fun putInPreferences(markerSet: Set<String>) {
        preferences.edit().clear().putStringSet(MARKERS_KEY, markerSet).apply()
    }

    private fun simpleMarkerToJson(marker: SimpleMarker): String? {
        val jsonObject = JSONObject()
        try {
            jsonObject.put(MARKER_TITLE_KEY, marker.title)
            jsonObject.put(MARKER_LATITUDE_KEY, marker.position.latitude)
            jsonObject.put(MARKER_LONGITUDE_KEY, marker.position.longitude)
            return jsonObject.toString()
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return null
    }

    private fun simpleMarkerFromJson(json: String): SimpleMarker? {
        try {
            val jsonObject = JSONObject(json)
            return SimpleMarker(
                    jsonObject.getString(MARKER_TITLE_KEY),
                    LatLng(jsonObject.getDouble(MARKER_LATITUDE_KEY), jsonObject.getDouble(MARKER_LONGITUDE_KEY))
            )
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return null
    }

}